File: src/main/java/com/cuupa/classificator/services/kb/semantic/text/SearchInternalLevenshtein.java
package com.cuupa.classificator.services.kb.semantic.text;

import org.apache.commons.text.similarity.LevenshteinDistance;

class SearchInternalLevenshtein {

    private final PlainText plainText;

    private SearchText wordsToSearch;

    private int tolerance;

    private int currentPositionPlainText;

    private int currentPositionSearchString;

    private int matchingWords;

    private boolean continueSearch;

    private int distance;

    SearchInternalLevenshtein(SearchText wordsToSearch, PlainText plaintText, int tolerance) {
        this.wordsToSearch = wordsToSearch;
        this.tolerance = tolerance;
        this.plainText = plaintText;
        continueSearch = true;
    }

    int getCurrentPositionPlainText() {
        return currentPositionPlainText;
    }

    int getCurrentPositionSearchString() {
        return currentPositionSearchString;
    }

    int getMatchingWords() {
        return matchingWords;
    }

    SearchInternalLevenshtein invoke() {
        String currentWordFromPlain = plainText.get(currentPositionPlainText);
        String currentWordToSearch = wordsToSearch.get(currentPositionSearchString);

        Integer distance = LevenshteinDistance.getDefaultInstance().apply(currentWordFromPlain, currentWordToSearch);

        if (distance <= tolerance) {
            matchingWords++;
            currentPositionPlainText++;
            currentPositionSearchString++;
            this.distance += distance;
        } else if (currentPositionSearchString > 0) {
            matchingWords = 0;
            currentPositionSearchString = 0;
            this.distance = 0;
            currentWordToSearch = wordsToSearch.get(currentPositionSearchString);
            distance = LevenshteinDistance.getDefaultInstance().apply(currentWordFromPlain, currentWordToSearch);

            if (distance <= tolerance) {
                matchingWords++;
                currentPositionSearchString++;
            }
            currentPositionPlainText++;
        } else {
            currentPositionPlainText++;
        }
        return this;
    }

    boolean isToSearch() {
        return continueSearch;
    }

    void cancelSearch() {
        continueSearch = false;
    }

    void resetCurrentPositionSearchString() {
        currentPositionSearchString = 0;
    }

    int getDistance() {
        return distance;
    }
}

File: src/main/java/com/cuupa/classificator/services/kb/semantic/text/SearchInternalLevenshtein.kt
package com.cuupa.classificator.services.kb.semantic.text

import org.apache.commons.text.similarity.LevenshteinDistance

internal class SearchInternalLevenshtein(private val wordsToSearch: SearchText, private val plainText: PlainText,
                                         private val tolerance: Int) {
    var currentPositionPlainText = 0
        private set
    var currentPositionSearchString = 0
        private set
    var matchingWords = 0
        private set
    var isToSearch = true
        private set
    var distance = 0
        private set

    operator fun invoke(): SearchInternalLevenshtein {
        val currentWordFromPlain = plainText[currentPositionPlainText]
        var currentWordToSearch = wordsToSearch[currentPositionSearchString]
        var distance = LevenshteinDistance.getDefaultInstance().apply(currentWordFromPlain, currentWordToSearch)
        when {
            distance <= tolerance -> {
                matchingWords++
                currentPositionPlainText++
                currentPositionSearchString++
                this.distance += distance
            }
            currentPositionSearchString > 0 -> {
                matchingWords = 0
                currentPositionSearchString = 0
                this.distance = 0
                currentWordToSearch = wordsToSearch[currentPositionSearchString]
                distance = LevenshteinDistance.getDefaultInstance().apply(currentWordFromPlain, currentWordToSearch)
                if (distance <= tolerance) {
                    matchingWords++
                    currentPositionSearchString++
                }
                currentPositionPlainText++
            }
            else -> {
                currentPositionPlainText++
            }
        }
        return this
    }

    fun cancelSearch() {
        isToSearch = false
    }

    fun resetCurrentPositionSearchString() {
        currentPositionSearchString = 0
    }

}